drivers/tee/tee_core.c

```diff
@@ -111,6 +121,209 @@ static int tee_release(struct inode *inode, struct file *filp)
 	return 0;
 }
 
+static const char *tee_session_get_application_id(void)
+{
+	return NULL;
+}
+
+static void tee_session_free_application_id(const char *app_id)
+{
+}
+
+int tee_session_verify_perm(u32 connection_method,
+			    const u8 connection_data[TEE_IOCTL_UUID_LEN])
+{
+	gid_t ns_grp = (gid_t)-1;
+	kgid_t grp = INVALID_GID;
+
+	/* Verify group memberships for logins including group */
+	if ((connection_method == TEE_IOCTL_LOGIN_GROUP) ||
+	    (connection_method == TEE_IOCTL_LOGIN_GROUP_APPLICATION)) {
+		memcpy(&ns_grp, connection_data, sizeof(gid_t));
+		grp = make_kgid(current_user_ns(), ns_grp);
+
+		if (!gid_valid(grp) || !in_egroup_p(grp))
+			return -EPERM;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(tee_session_verify_perm);
+
+static int uuid_v5(uuid_t *uuid, const uuid_t *ns, const void *name,
+		   size_t size)
+{
+	unsigned char hash[SHA1_DIGEST_SIZE];
+	struct crypto_shash *shash = NULL;
+	struct shash_desc *desc = NULL;
+	int rc;
+
+	shash = crypto_alloc_shash("sha1", 0, 0);
+	if (IS_ERR(shash)) {
+		rc = PTR_ERR(shash);
+		pr_err("shash(sha1) allocation failed\n");
+		return rc;
+	}
+
+	desc = kzalloc(sizeof(*desc) + crypto_shash_descsize(shash),
+		       GFP_KERNEL);
+	if (IS_ERR(desc)) {
+		rc = PTR_ERR(desc);
+		goto out;
+	}
+
+	desc->tfm = shash;
+
+	rc = crypto_shash_init(desc);
+	if (unlikely(rc < 0))
+		goto out2;
+
+	rc = crypto_shash_update(desc, (const u8 *)ns, sizeof(*ns));
+	if (unlikely(rc < 0))
+		goto out2;
+
+	rc = crypto_shash_update(desc, (const u8 *)name, size);
+	if (unlikely(rc < 0))
+		goto out2;
+
+	rc = crypto_shash_final(desc, hash);
+	if (unlikely(rc < 0))
+		goto out2;
+
+	// time_low
+	uuid->b[0] = hash[0];
+	uuid->b[1] = hash[1];
+	uuid->b[2] = hash[2];
+	uuid->b[3] = hash[3];
+
+	// time_mid
+	uuid->b[4] = hash[4];
+	uuid->b[5] = hash[5];
+
+	// time_hi_and_version (v5)
+	uuid->b[6] = (hash[6] & 0x0F) | 0x50;
+	uuid->b[7] = hash[7];
+
+	// clock_seq_hi_and_reserved (variant)
+	uuid->b[8] = (hash[8] & 0x3F) | 0x80;
+
+	// clock_seq_low
+	uuid->b[9] = hash[9];
+
+	// node
+	uuid->b[10] = hash[10];
+	uuid->b[11] = hash[11];
+	uuid->b[12] = hash[12];
+	uuid->b[13] = hash[13];
+	uuid->b[14] = hash[14];
+	uuid->b[15] = hash[15];
+
+out2:
+	kfree(desc);
+
+out:
+	crypto_free_shash(shash);
+	return rc;
+}
+
+int tee_session_calc_client_uuid(uuid_t *uuid, u32 connection_method,
+				 const u8 connection_data[TEE_IOCTL_UUID_LEN])
+{
+	const char *application_id = NULL;
+	gid_t ns_grp = (gid_t)-1;
+	kgid_t grp = INVALID_GID;
+	char *name = NULL;
+	int rc;
+
+	if (connection_method == TEE_IOCTL_LOGIN_PUBLIC) {
+		/* Nil UUID to be passed to TEE environment */
+		uuid_copy(uuid, &uuid_null);
+		return 0;
+	}
+
+	/*
+	 * In Linux environment client UUID is based on UUIDv5.
```

Comment:
Do you have a link/reference to some standard or convention?

Reply:
Probably the easiest is to have a look at wikipedia:
https://en.wikipedia.org/wiki/Universally_unique_identifier

And then the actual RFC:
https://www.rfc-editor.org/rfc/rfc4122.txt

If you talk about the "Linux environment" part this is an agreement that needs to be made. GlobalPlatform leaves this as undefined and it is up to REE implementation to decide how to do it.

But in the end it must be UUID and only deterministic choices are then UUIDv3 or UUIDv5 where UUIDv5 is slightly better.

Reply:
Yeah, it was rather that GP leaves this undefined and we're improvising or proposing a way to deal with it. Boot loaders may access TAs too, so we need something that can be replicated there if needed.

Reply:
In this case I would say there needs to be unique namespace UUID allocated for that case. Perhaps this should be managed in OP-TEE as a registry or so.

Reply:
> Perhaps this should be managed in OP-TEE as a registry or so.

If needed later on we'll do something about that. What's important here is to be clear that GP leaves this undefined and we're filling the gap in this way.